### PR TITLE
[ENG-3865] add support for uploading build artifacts

### PR DESCRIPTION
**Pull request by dsokal**
# Checklist

- [ ] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

https://linear.app/expo/issue/ENG-3865/testing-on-eas-build-documentation

We want to enable users to upload any arbitrary build artifacts + application archive from EAS Build. The use case is so that users can upload screenshots of failed E2E tests.

# How

- Add new fields to eas.json: `applicationArchivePath` and `buildArtifactsPaths`.
- Deprecate `artifactPath`. EAS CLI offers to automatically rename the deprecated field.
- Update `eas.json` schema.

# Test Plan

Tested with https://github.com/expo/eas-tests-example

Rewriting eas.json fields:
Screenshot 2022-08-31 at 11 24 48

### CODE REVIEW

File: packages/eas-json/schema/eas.schema.json [hunk 1]
```diff
@@ -168,6 +168,14 @@
           "markdownDescription": "When enabled, for iOS, bumps the last component of `expo.ios.buildNumber` (e.g. `1.2.3.39` -> `1.2.3.40`) and for Android, bumps `expo.android.versionCode` (e.g. `3` -> `4`).",
           "default": false
         },
+        "buildArtifactsPaths": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
```
Comment (dsokal) [inline packages/eas-json/schema/eas.schema.json:175]: @byCedric is this the correct way to define an array of strings? How can I test it?

File: packages/eas-cli/src/commands/build/index.ts [hunk 2]
```diff
@@ -208,3 +217,71 @@ export default class Build extends EasCommand {
...
+  const profileNames = await reader.getBuildProfileNamesAsync();
+  const platformAndProfileNames: [Platform, string][] = profileNames.flatMap(profileName => [
+    [Platform.ANDROID, profileName],
+    [Platform.IOS, profileName],
+  ]);
+
+  const deprecatedProfiles: [Platform, string][] = [];
+
+  for (const [platform, profileName] of platformAndProfileNames) {
+    const buildProfile = await reader.getBuildProfileAsync(platform, profileName);
```
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/index.ts:235]: consider reading just raw json, with this change if someone has an invalid profile in eas.json it won't be possible to build for any other profile too

File: packages/eas-json/schema/eas.schema.json [hunk 3]
```diff
@@ -244,10 +252,10 @@
           "description": "Gradle task that will be used to build your project, e.g. `:app:assembleDebug` to build a debug binary.",
           "markdownDescription": "Gradle task that will be used to build your project, e.g. `:app:assembleDebug` to build a debug binary.\n\nIt's not recommended unless you need to run a task that `buildType` does not support, it takes priority over `buildType` and `developmentClient`."
         },
-        "artifactPath": {
+        "applicationArchivePath": {
```
Comment (wkozyra95) [inline packages/eas-json/schema/eas.schema.json:255]: I do not have an opinion on that, but this schema is not distributed with eas-cli, so after publishing artfiactPath will be marked as error/unknown property in vscode, even if someone is on older eas-cli that does still support it.

File: packages/eas-json/schema/eas.schema.json (hunk 1 above)
Comment (wkozyra95) [reply packages/eas-json/schema/eas.schema.json:175]: > is this the correct way to define an array of strings?
yes, it works for mee

Not sure how to test it in vscode, I'm testing that by changing URL to the schema in my vim config and opening eas.json.

You could probably use sth like this https://www.jsonschemavalidator.net/

File: packages/eas-json/src/build/schema.ts [hunk 4]
```diff
@@ -39,10 +40,12 @@ const AndroidBuildProfileSchema = CommonBuildProfileSchema.concat(
     ),
 
     artifactPath: Joi.string(),
+    applicationArchivePath: Joi.string(),
+    buildArtifactsPaths: Joi.array().items(Joi.string()),
```
Comment (wkozyra95) [inline packages/eas-json/src/build/schema.ts:44]: I think you originally wanted to make it `string | string[]`, Did you change your mind?

**Review by wkozyra95 — APPROVED** [review]

File: packages/eas-json/src/build/schema.ts (hunk 4 above)
Comment (dsokal) [reply packages/eas-json/src/build/schema.ts:44]: @brentvatne 
Screenshot 2022-08-31 at 14 48 04

**Review by szdziedzic — APPROVED** [review]

**Review by brentvatne — APPROVED** [review]

File: packages/eas-cli/src/commands/build/index.ts [hunk 5]
```diff
@@ -208,3 +217,71 @@ export default class Build extends EasCommand {
...
+    Log.warn(
+      `See ${link('https://docs.expo.dev/build-reference/eas-json/')} for more information.`
+    );
+    Log.warn(
+      'This warning will become an error in an upcoming EAS CLI release. For now, we will proceed to avoid disruption of your builds.'
+    );
+    return;
+  }
+
+  const rename = await selectAsync('Do you want to handle renaming the field for you?', [
```
Comment (kbrandwijk) [inline packages/eas-cli/src/commands/build/index.ts:267]:
```suggestion
  const rename = await selectAsync('Do you want EAS CLI to handle renaming the field for you?', [
```
Or something similar.

File: packages/eas-cli/src/commands/build/index.ts [hunk 6]
```diff
@@ -208,3 +217,71 @@ export default class Build extends EasCommand {
...
+      deprecatedProfiles.push([platform, profileName]);
+    }
+  }
+
+  if (deprecatedProfiles.length === 0) {
+    return;
+  }
+
+  const deprecatedProfileNames = uniq(deprecatedProfiles.map(([, profileName]) => profileName));
+  Log.warn(`Some of your build profiles use deprecated ${chalk.bold('artifactPath')} field:`);
```
Comment (kbrandwijk) [inline packages/eas-cli/src/commands/build/index.ts:246]:
```suggestion
  Log.warn(`Some of your build profiles use deprecated field ${chalk.bold('artifactPath')}:`);
```

File: packages/eas-cli/src/commands/build/index.ts [hunk 7]
```diff
@@ -208,3 +217,71 @@ export default class Build extends EasCommand {
...
+    Log.warn(
+      `${figures.warning} Action required: rename ${chalk.bold('artifactPath')} to ${chalk.bold(
+        'applicationArchivePath'
+      )} in all of the build profiles listed above.`
+    );
+    Log.warn(
+      `See ${link('https://docs.expo.dev/build-reference/eas-json/')} for more information.`
+    );
+    Log.warn(
+      'This warning will become an error in an upcoming EAS CLI release. For now, we will proceed to avoid disruption of your builds.'
```
Comment (kbrandwijk) [inline packages/eas-cli/src/commands/build/index.ts:262]:
```suggestion
      'This warning will become an error in a future EAS CLI release. This build will continue to use the ${chalk.bold('artifactPath')} setting.'
```

File: packages/eas-cli/src/commands/build/index.ts (hunk 5 above)
Comment (kbrandwijk) [reply packages/eas-cli/src/commands/build/index.ts:267]:
```suggestion
  const rename = await selectAsync('Do you want us to handle renaming the field for you?', [
```

File: packages/eas-cli/src/commands/build/index.ts [hunk 8]
```diff
@@ -208,3 +217,71 @@ export default class Build extends EasCommand {
...
+    );
+    return;
+  }
+
+  const rename = await selectAsync('Do you want to handle renaming the field for you?', [
+    { title: 'Yes', value: true },
+    { title: 'No, I will edit eas.json on my own (EAS CLI exits)', value: false },
```
Comment (kbrandwijk) [inline packages/eas-cli/src/commands/build/index.ts:269]:
```suggestion
    { title: 'No, I will edit eas.json manually (EAS CLI exits)', value: false },
```

File: packages/eas-json/schema/eas.schema.json [hunk 9]
```diff
@@ -168,6 +168,14 @@
           "markdownDescription": "When enabled, for iOS, bumps the last component of `expo.ios.buildNumber` (e.g. `1.2.3.39` -> `1.2.3.40`) and for Android, bumps `expo.android.versionCode` (e.g. `3` -> `4`).",
           "default": false
         },
+        "buildArtifactPaths": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "List of paths (or patterns) where EAS Build is going to look for the build artifacts. Use `applicationArchivePath` for specifying the path for uploading the application archive. Build artifacts are uploaded even if the build fails. See: https://github.com/mrmlnc/fast-glob#pattern-syntax",
```
Comment (kbrandwijk) [inline packages/eas-json/schema/eas.schema.json:176]:
```suggestion
          "description": "List of paths (or patterns) where EAS Build is going to look for the build artifacts. Use `applicationArchivePath` to specify the path for uploading the application archive. Build artifacts are uploaded even if the build fails. See: https://github.com/mrmlnc/fast-glob#pattern-syntax",
```

File: packages/eas-json/schema/eas.schema.json [hunk 10]
```diff
@@ -168,6 +168,14 @@
...
+          "markdownDescription": "List of paths (or patterns) where EAS Build is going to look for the build artifacts. Use `applicationArchivePath` for specifying the path for uploading the application archive. Build artifacts are uploaded even if the build fails. EAS Build uses the `fast-glob` npm package for pattern matching ([see their README to learn more about the syntax you can use](https://github.com/mrmlnc/fast-glob#pattern-syntax))."
```
Comment (kbrandwijk) [inline packages/eas-json/schema/eas.schema.json:177]:
```suggestion
          "markdownDescription": "List of paths (or patterns) where EAS Build is going to look for the build artifacts. Use `applicationArchivePath` to specify the path for uploading the application archive. Build artifacts are uploaded even if the build fails. EAS Build uses the `fast-glob` npm package for pattern matching ([see their README to learn more about the syntax you can use](https://github.com/mrmlnc/fast-glob#pattern-syntax))."
```

**Review by kbrandwijk — APPROVED** [review]
Some minor copy suggestions.

File: packages/eas-json/schema/eas.schema.json (hunk 3 above)
Comment (dsokal) [reply packages/eas-json/schema/eas.schema.json:255]: I added `artifactPath` back and marked as deprecated.

File: packages/eas-cli/src/commands/build/index.ts (hunk 2 above)
Comment (dsokal) [reply packages/eas-cli/src/commands/build/index.ts:235]: Acknowledged, won't fix :D

**Comment by dsokal** [comment]
/changelog-entry new-feature Add support for `buildArtifactPaths`. Rename `artifactPath` to `applicationArchivePath`.

**Pull request merged**